llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp

```diff
@@ -80,6 +80,10 @@ enum class SchedGroupMask {
   LLVM_MARK_AS_BITMASK_ENUM(/* LargestFlag = */ ALL)
 };
 
+static bool handleAsVMEMInstr(const MachineInstr &MI, const SIInstrInfo *TII) {
+  return TII->isVMEM(MI) || (TII->isFLAT(MI) && !TII->isDS(MI));
```

Comment:
I don't understand this condition. isVMEM should cover it. There should also be no instructions that are flat and DS

Reply:
If I only use `TII->isVMEM(MI)` in all places I currently use `handleAsVMEMInstr`, there are some failing tests:
```
  LLVM :: CodeGen/AMDGPU/llvm.amdgcn.sched.group.barrier.ll
  LLVM :: CodeGen/AMDGPU/sched-barrier-hang-weak-dep.mir
  LLVM :: CodeGen/AMDGPU/sched-barrier-pre-RA.mir
  LLVM :: CodeGen/AMDGPU/sched-group-barrier-pipeline-solver.mir
  LLVM :: CodeGen/AMDGPU/sched-group-barrier-pre-RA.mir
```
There seems to be `GLOBAL_STORE_DWORDX4_SADDR %14:vgpr_32, %135:vreg_128_align2, %143.sub2_sub3:sgpr_128, 0, 0, implicit $exec :: (store (s128) into %ir.gep2, align 128, addrspace 1)`, for example, which is both FLAT and !DS

Reply:
Yes, it is a FLAT encoded instruction and not a DS encoded instruction. Both of those are trivially true. But it should be covered by isVMEM

Reply:
Apparently not:
```
(gdb) p MI.dump()
  GLOBAL_STORE_DWORDX4_SADDR %14:vgpr_32, %135:vreg_128_align2, %143.sub2_sub3:sgpr_128, 0, 0, implicit $exec :: (store (s128) into %ir.gep2, align 128, addrspace 1)
$1 = void
(gdb) p TII->isVMEM(MI)
$2 = false
(gdb) p TII->isFLAT(MI)
$3 = true
(gdb) p TII->isDS(MI)
$4 = false
```

Reply:
Opened #137148

Reply:
(Created https://github.com/llvm/llvm-project/pull/137148 and https://github.com/llvm/llvm-project/pull/137170)